### DISCUSSION TITLE
Welcome to fprime-system-reference Discussions!

### DISCUSSION THREAD

**Discussion by LeStarch** [fprime-community/fprime-system-reference#27] [Announcements]
Welcome. This discussions forum is a place to ask questions about the fprime software class setup.

**Comment by ksali2** [comment]
Using my JPL Windows laptop, I am unable to install the python3-venv and python3-pip packages.  I get the following error messages indicating that those packages are unavailable.

```
ksali@MT-305367:~$ sudo apt install build-essential git g++ gdb cmake python3 python3-venv python3-pip
[sudo] password for ksali:
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package python3-pip is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

Package python3-venv is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'python3-venv' has no installation candidate
E: Package 'python3-pip' has no installation candidate
```

![image](https://user-images.githubusercontent.com/115734276/195667940-7c6bcd8b-296f-4c9d-b8f1-a57150b5a42d.png)

**Reply by jcarste** [reply]
Doing this first seemed to resolve it for me:
sudo apt-get update

**Reply by LeStarch** [reply]
Good catch!  Yes, `sudo apt update` will ensure package listings are up-to-date.

**Reply by LeStarch** [reply]
I have added that to the list of instructions

**Reply by ksali2** [reply]
Thanks!  That got me past that part of the instructions.

**Comment by djbyrne17** [comment]
"git submodule update --init" lacks " --recursive" in [prerequisites](https://github.com/fprime-community/fprime-system-reference/blob/main/docs/course/prerequisites.md#step-2:-cloning-the-f%C2%B4-system-reference) Step 2, but has it in [appendix-3](https://github.com/fprime-community/fprime-system-reference/blob/main/docs/course/appendix-3.md#cmake-error-at-cmakeliststxt-line-32).  Which is correct?

**Reply by LeStarch** [reply]
```
git submodule update --init --recursive
```
Is correct, although in practice both will work as we only have 1-depth of submodules.

**Reply by LeStarch** [reply]
I have fixed the prerequsites.  Thanks for pointing that out.

**Reply by djbyrne17** [reply]
Thanks.  Also, the prerequisites state clearly that "**cross-compilers require a Linux computer**" and "**a Linux box is essential**" yet also point Mac users to appendix 2 for docker setup.  Is the intention that a Mac with docker enables cross-compilation, and is an alternative to a linux box?

**Reply by djbyrne17** [reply]
...in which case, prerequistes step 2.5 could be explicitly "start docker".

appendix-2.md, at the end, instead of the note "~/fprime-system-reference should be replaced with /project in commands run in the container." could simply have
 
> ln  -s /project ~/fprime-system-reference

**Reply by LeStarch** [reply]
Docker provides the Linux environment for macOS users like WSL does for Windows users.  I will add fixes for the above suggestions.

**Reply by LeStarch** [reply]
Looking into this deeper: Docker is only required for the cross-compile and testing on hardware section.  (Day 3 of the class).  All other days can be done on macOS directly.  I believe day 3 mentions running docker, but I will state this verbally that day.

**Reply by LeStarch** [reply]
Ok I added a note about the symbolic link. For the next class we can add that into the container so that it "just works" but that change is a little too destructive now that students are running through the exercises.

**Comment by nihaldhamani** [comment]
Running on M1 Mac within the docker cross-compile container. The `generate` and `build` commands work fine but I get this error when launching `fprime-gds`

```
(class-venv) fprime@lima-rancher-desktop:/project/SystemReference$ fprime-gds
[INFO] Autodetected artifacts root '/project/SystemReference/build-artifacts' from deployment settings.ini file.
[INFO] Ensuring TCP Server is stable for at least 5 seconds
[INFO] Running Application: /home/fprime/class-venv/bin/python3
[INFO] Log File: /project/SystemReference/logs/2022_10_14-18_22_32/ThreadedTCP.log
[INFO] Ensuring comm[ip] Application is stable for at least 1 seconds
[INFO] Running Application: /home/fprime/class-venv/bin/python3
[INFO] Ensuring SystemReference Application is stable for at least 1 seconds
[INFO] Running Application: /project/SystemReference/build-artifacts/Linux/bin/SystemReference
[INFO] Log File: /project/SystemReference/logs/2022_10_14-18_22_32/SystemReference.log
[ERROR] Failed to run application: /project/SystemReference/build-artifacts/Linux/bin/SystemReference -p 50000 -a 0.0.0.0. Error: /project/SystemReference/build-artifacts/Linux/bin/SystemReference stopped with code -6 sooner than 1 seconds.
    [LOG] Assert file "/project/fprime/Os/Posix/Mutex.cpp": Line: 3095
    [LOG] SystemReference: /project/fprime/Fw/Types/Assert.cpp:163: NATIVE_INT_TYPE Fw::SwAssert(const CHAR*, NATIVE_UINT_TYPE, FwAssertArgType): Assertion `0' failed.
[INFO] Shutting down F prime due to error. Failed to run SystemReference Application
```

**Reply by djbyrne17** [reply]
When I'd built all the code natively on Mac, then later entered docker, the make system did not see a need to re-compile inside docker and I got weird error msgs.

Try re-naming, or removing, /fprime-system-reference/SystemReference/build-fprime-automatic-native/
and repeating

> fprime-util generate
> fprime-util build
> fprime-gds

Doing that, I get

> (class-venv) DJ-Rdocker% fprime-gds
> [INFO] Autodetected artifacts root '/project/SystemReference/build-artifacts' from deployment settings.ini file.
> [INFO] Ensuring TCP Server is stable for at least 5 seconds
> [INFO] Running Application: /project/class-venv/bin/python3
> [INFO] Log File: /project/SystemReference/logs/2022_10_14-20_25_53/ThreadedTCP.log
> [INFO] Ensuring comm[ip] Application is stable for at least 1 seconds
> [INFO] Running Application: /project/class-venv/bin/python3
> [INFO] Ensuring SystemReference Application is stable for at least 1 seconds
> [INFO] Running Application: /project/SystemReference/build-artifacts/Linux/bin/SystemReference
> [INFO] Log File: /project/SystemReference/logs/2022_10_14-20_25_53/SystemReference.log
> [INFO] Ensuring HTML GUI is stable for at least 2 seconds
> [INFO] Running Application: /project/class-venv/bin/python3
>  * Serving Flask app "fprime_gds.flask.app"
>  * Environment: production
>    WARNING: This is a development server. Do not use it in a production deployment.
>    Use a production WSGI server instead.
>  * Debug mode: off
> [INFO] F prime is now running. CTRL-C to shutdown all components.

However, I do not get the GDS GUI in a web browser.  Perhaps Rancher docker needs some port-forwarding to let 127.0.0.1:50000 get out of docker and into my may firefox.  But we can get by without that, since we can code and debug in Mac (and write solid Unit Tests, of course!), and only use docker to cross-compile for the HW target on the last day of class.  I think.

**Reply by djbyrne17** [reply]
My Mac is Intel, not M1 like yours.

- 2.4 GHz 8-Core Intel Core i9
- Monterey 12.6

**Reply by LeStarch** [reply]
This is correct.  We typically run only the cross-compile build in the container.  We build and develop for the local development and tests natively on macOS.  Thanks @djbyrne17.

If you want to really want to run inside docker you can repeat the following steps (inside the docker container) to generate a native linux build as the container runs docker.

```
fprime-util purge 
fprime-util generate
fprime-util build
fprime-gds
```

To see the GDS you'll need to open a browser and manually go to `http://localhost:5000` directly.  The ports are forwarded but the command to "launch a browser" cannot escape the container.

**Reply by nihaldhamani** [reply]
Ah I see, thanks! @djbyrne17 @LeStarch

**Comment by nadyabalabanska** [comment]
In the F' GDS window, I am not seeing the upper right corner change to a green circle. I tried sending a command, but there still remained only a red X.

**Reply by LeStarch** [reply]
Is this WSL, macOS, or Linux?  On a local machine (e.g. laptop)?  What errors/logs do you see (CLI output, "Logs" tab of the GDS)?

We can try to debug here.  There will also be some time Monday to help with issues that came up while people installed.

**Reply by nadyabalabanska** [reply]
I am using WSL on my JPL Windows laptop. When I run `fprime-gds` a Chrome window opens connected to 127.0.0.1:5000 which looks like this 
![Screenshot (120)](https://user-images.githubusercontent.com/16469161/195945583-3df1d32b-5d1f-48da-ba4d-e3d3885bf53d.png)

I am not seeing the green circle in the upper right hand corner, just the red X. 

No error output in the terminal: 
```
[INFO] Autodetected artifacts root '/home/nadyaba/fprime-system-reference/SystemReference/build-artifacts' from deployment settings.ini file.
[INFO] Ensuring TCP Server is stable for at least 5 seconds 
[INFO] Running Application: /home/nadyaba/class-venv/bin/python3
[INFO] Log File: /home/nadyaba/fprime-system-reference/SystemReference/logs/2022_10_14-14_07_28/ThreadedTCP.log 
[INFO] Ensuring comm[ip] Application is stable for at least 1 seconds
[INFO] Running Application: /home/nadyaba/class-venv/bin/python3
[INFO] Ensuring SystemReference Application is stable for at least 1 seconds
[INFO] Running Application: /home/nadyaba/fprime-system-reference/SystemReference/build-artifacts/Linux/bin/SystemReference
[INFO] Log File: /home/nadyaba/fprime-system-reference/SystemReference/logs/2022_10_14-14_07_28/SystemReference.log
[INFO] Ensuring HTML GUI is stable for at least 2 seconds
[INFO] Running Application: /home/nadyaba/class-venv/bin/python3
* Serving Flask app "fprime_gds.flask.app"
* Environment: production
WARNING: This is a development server. Do not use it in a production deployment.
Use a production WSGI server instead.
 * Debug mode: off
 [INFO] F prime is now running. CTRL-C to shutdown all components.  
```

Nothing in the Logs tab of GDS either.

**Reply by LeStarch** [reply]
We may need to look at this tomorrow.  It seems to be running, but the connections aren't working.  Are you running any VPN fixes or anything like that for WSL?

**Comment by albee** [comment]
It looks like [fprime-fpp](https://github.com/fprime-community/fpp/releases/tag/v1.0.2) is missing an ARM64 release for those on Mac M1's (see below). I happen to also have an x86 laptop that I've been able to set up F' on, but I know a few coworkers were also having the same issue running Ubuntu 20.04 via Parallels on Mac M1. Is there any suggested workaround?

```
Collecting fprime-fpp==1.0.2
  Using cached fprime-fpp-1.0.2.tar.gz (615 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 255
  ╰─> [9 lines of output]
      running egg_info
      creating /tmp/pip-pip-egg-info-qqotf5pt/fprime_fpp.egg-info
      writing /tmp/pip-pip-egg-info-qqotf5pt/fprime_fpp.egg-info/PKG-INFO
      writing dependency_links to /tmp/pip-pip-egg-info-qqotf5pt/fprime_fpp.egg-info/dependency_links.txt
      writing top-level names to /tmp/pip-pip-egg-info-qqotf5pt/fprime_fpp.egg-info/top_level.txt
      writing manifest file '/tmp/pip-pip-egg-info-qqotf5pt/fprime_fpp.egg-info/SOURCES.txt'
      -- INFO  -- Fetching FPP tools at https://github.com/fprime-community/fpp/releases/download/v1.0.2/native-fpp-Linux-aarch64.tar.gz
      -- WARN  -- Failed to retrieve https://github.com/fprime-community/fpp/releases/download/v1.0.2/native-fpp-Linux-aarch64.tar.gz with error: HTTP Error 404: Not Found
      -- ERROR -- java must exist on PATH to build from source
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed
```

**Reply by LeStarch** [reply]
Yes, we do not currently have support for any of the arm variants.  This is why the mac OS instructions recommend docker as opposed to parallels or other virtualization technology and running on ubuntu arm64.

**Comment by jamred** [comment]
Following along with [appendix 2](https://github.com/fprime-community/fprime-system-reference/blob/main/docs/course/appendix-2.md) on an M1 mac, when I run the container I get the following warning: `WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested`.

I don't think this is actually a problem, and it is running the container, but I just wanted to document it.

**Reply by LeStarch** [reply]
Thanks for posting.  I don't think this is an issue either.  This will help others who come to post on that issue.
